### PR TITLE
#657 - Properly find libruby/libpythonengine on macOS

### LINKED ISSUE #657: Can't run Python measure on Mac M1
Issue overview
--------------
OpenStudio CLI looking for libopenstudiolib.dylib in incorrect location

## Current Behavior

```
stderr: Traceback (most recent call last):
stderr: File "/Applications/OpenStudioApplication-1.7.0-rc1/bin/../Python/openstudio.py", line 47, in <module>
stderr: import openstudioairflow as airflow
stderr: File "/Applications/OpenStudioApplication-1.7.0-rc1/bin/../Python/openstudioairflow.py", line 12, in <module>
stderr: import _openstudioairflow
stderr: ImportError: dlopen(/Applications/OpenStudioApplication-1.7.0-rc1/bin/../Python/_openstudioairflow.so, 0x0004): Library not loaded: @rpath/libopenstudiolib.dylib
stderr: Referenced from: <2EF0357C-4D3E-3F51-A3F5-008565FD679B> /Applications/OpenStudioApplication-1.7.0-rc1/Python/_openstudioairflow.so
stderr: Reason: tried: '/Applications/OpenStudioApplication-1.7.0-rc1/Python/libopenstudiolib.dylib' (no such file), '/Applications/OpenStudioApplication-1.7.0-rc1/Python/../lib/libopenstudiolib.dylib' (no such file), '/Applications/OpenStudioApplication-1.7.0-rc1/Python/libopenstudiolib.dylib' (no such file), '/Applications/OpenStudioApplication-1.7.0-rc1/Python/../lib/libopenstudiolib.dylib' (no such file), '/opt/homebrew/lib/libopenstudiolib.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libopenstudiolib.dylib' (no such file), '/opt/homebrew/lib/libopenstudiolib.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libopenstudiolib.dylib' (no such file), '/opt/homebrew/lib/libopenstudiolib.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libopenstudiolib.dylib' (no such file)
[openstudio.workflow.OSWorkflow] <Error> Found error in state 'OpenStudioMeasures' with message: 'Error executing Python code'
Timer OpenStudioMeasures with level=0 was not stopped
stderr: Failed to run workflow. Last Error:
stderr: Found error in state 'OpenStudioMeasures' with message: 'Error executing Python code'
```

## Expected Behavior
CLI should look for libopenstudiolib.dylib in correct location and run successfully

## Steps to Reproduce

1. Create Python measure
2. Add Python measure to workflow
3. Run workflow

## Possible Solution
Either change rpath to include ./bin/ or install libopenstudiolib.dylib to ./lib/.  Copying libopenstudiolib.dylib to ./lib/ is a temporary workaround

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): M1 MacOS 13.5
 - Version of OpenStudioApplication (if using an intermediate build, include SHA): 1.7.0-rc1

### Context
Cannot use Python measures

### PR DESCRIPTION
**Pull request by jmarrec**
* Fix #657 

Adjust location of libruby/libpythonengine to root/lib (for macOS mostly).

I have tested on mac that this works in the build dir + the install dir **BUT** on my same machine. I need to redo that testing with the package produced on CI to make sure it works.

Testing on linux and windows is needed too.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -728,9 +728,9 @@ install(DIRECTORY "${openstudio_ROOT_DIR}/Python" DESTINATION "." COMPONENT "Pyt
 # install(DIRECTORY "${openstudio_ROOT_DIR}/lib" DESTINATION "." COMPONENT "OpenStudioApp" USE_SOURCE_PERMISSIONS)
 
 install(IMPORTED_RUNTIME_ARTIFACTS openstudio::openstudio DESTINATION bin COMPONENT "CLI")
-install(IMPORTED_RUNTIME_ARTIFACTS openstudio::openstudiolib DESTINATION bin COMPONENT "CLI")
-install(IMPORTED_RUNTIME_ARTIFACTS openstudio::rubyengine DESTINATION bin COMPONENT "CLI")
-install(IMPORTED_RUNTIME_ARTIFACTS openstudio::pythonengine DESTINATION bin COMPONENT "CLI")
+install(IMPORTED_RUNTIME_ARTIFACTS openstudio::openstudiolib DESTINATION lib COMPONENT "CLI")
```
Comment (macumber) [inline CMakeLists.txt:731]: I think we want these in bin on Windows but lib on Linux/Mac

**Comment by macumber** [comment]
This should fix https://github.com/openstudiocoalition/OpenStudioApplication/issues/657 right?

Not sure how hard it would be but it would be awesome to install the installer in the CI and then use it to run a workflow using both Ruby and Python measures.  Maybe something for next release?

File: CMakeLists.txt (hunk 1 above)
Comment (macumber) [reply CMakeLists.txt:731]: I downloaded the installer and confirmed this should be bin on Windows.  The CLI wasn't working directly from the install.  I moved these three libs from the lib dir to bin dir and it worked in both the OS App and SketchUp

**Comment by jmarrec** [comment]
@macumber Would we do that on each run of app_build.yml? Or a manual trigger workflow (maybe taking a run_id as an input so we locate the artifacts that were uploaded)?

**Comment by macumber** [comment]
The test workflow could probably be manually triggered, it would be nice to work out the kinks in the automatic install scripts too, I added https://github.com/openstudiocoalition/OpenStudioApplication/issues/665 to track it

**Comment by jmarrec** [comment]
I've start this branch: https://github.com/openstudiocoalition/OpenStudioApplication/tree/cli_test

getting some problems setting it up / some failures during testing. I'm currently investigating

**Comment by jmarrec** [comment]
I have some issues based on the testing.

On linux there is both `lib/libopenstudiolib.so`and `bin/libopenstudiolib.so`. The bin/ one should NOT exist because that's what's found by the CLI, and it assumes the librubyengine/libpythoinengine is next to it.
the libopenstudiolib has an RPATH of that checks $ORIGIN and $ORIGIN/../lib.

I'm not sure what places the libopenstudiolib.so. in bin/ but I think it's `fixup_bundle` which is like a black block to me

**Comment by jmarrec** [comment]
Attempted a fix on [aab5fa4](https://github.com/openstudiocoalition/OpenStudioApplication/pull/664/commits/aab5fa44fa3dc366ab185f7d0816ae845782ea89). I've got hope. Local testing seemed fine on linux and mac. Will wait for CI here to finish, then I'll try the cli_test workflow again.

On macos, this eliminates the duplicated libopenstudiolib.dylib (we used to have 3...) and the duplicated openstudio (exe, we used to have two)


At some point we should really fix the mac bundle... it's a gigantic mess, having stuff both in OpenStudioApp.app/ and at the root.

For now on mac, the CLI is **only** at OpenStudio.app/Contents/MacOS/openstudio. Which looks for libopenstudiolib in Contents/Frameworks/, which itself has symlinks to the root lib/ dir.

```
lrwxr-xr-x  1 julien  staff  35 Jan  9 17:45 ./OpenStudioApp.app/Contents/Frameworks/libopenstudiolib.dylib -> ../../../lib/libopenstudiolib.dylib
lrwxr-xr-x  1 julien  staff  31 Jan  9 17:45 ./OpenStudioApp.app/Contents/Frameworks/libpythonengine.so -> ../../../lib/libpythonengine.so
lrwxr-xr-x  1 julien  staff  29 Jan  9 17:45 ./OpenStudioApp.app/Contents/Frameworks/librubyengine.so -> ../../../lib/librubyengine.so
```

That way the CLI works, and the Python/Ruby bindings which are at root can find the ../lib/ ones too.

**Pull request merged**